File: protocol/src/interfaces/Room.ts
export interface ServerSend {
    startingIn: (n: number) => void
    starting: () => void
    state: (state: DetailedRoomState) => void
}

export interface ClientSend {
    register: (registration: PlayerRegistration) => boolean
    leave: () => void
    ready: (ready: boolean) => void,
    getState: () => DetailedRoomState;
}

export interface PlayerRegistration {
    name: string,
    color: string
}

export interface Player {
    name: string,
    color: string,
    ready: boolean
};

export interface DetailedRoomState {
    id: string,
    current: number,
    name: string,
    max: number,
    players: Player[],
    ingame: boolean
}

File: mobile/lib/external/protocol/interfaces/room.dart
import 'package:json_annotation/json_annotation.dart';

part 'room.g.dart';

@JsonSerializable()
class PlayerRegistration {
  PlayerRegistration({required this.name, required this.color});

  String name;
  String color;

  factory PlayerRegistration.fromJson(Map<String, dynamic> json) => _$PlayerRegistrationFromJson(json);
  Map<String, dynamic> toJson() => _$PlayerRegistrationToJson(this);
}

@JsonSerializable()
class Player {
  Player({required this.name, required this.color, required this.ready});

  String name;
  String color;
  bool ready;

  factory Player.fromJson(Map<String, dynamic> json) => _$PlayerFromJson(json);
  Map<String, dynamic> toJson() => _$PlayerToJson(this);
}

@JsonSerializable()
class DetailedRoomState {
  DetailedRoomState({required this.id, required this.current, required this.name, required this.max, required this.players, required this.ingame});

  String id;
  int current;
  String name;
  int max;
  List<Player> players;
  bool ingame;

  factory DetailedRoomState.fromJson(Map<String, dynamic> json) => _$DetailedRoomStateFromJson(json);
  Map<String, dynamic> toJson() => _$DetailedRoomStateToJson(this);
}